### PR TITLE
fix todo:add function getFailContainer to report which containers failed the pod

### PR DESCRIPTION
**Pull request by hanxiaoshuai**
**What this PR does / why we need it**:
fix todo:add function getFailContainer to report which containers failed the pod in runonce.go
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by hanxiaoshuai** [comment]
cc @proppy @hzxuzhonghu

**Comment by hanxiaoshuai** [comment]
cc @yujuhong @tallclair

File: pkg/kubelet/runonce.go [hunk 1]
```diff
@@ -156,3 +156,18 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
...
+
+// getFailContainer returns failed container name for pod.
+func (kl *Kubelet) getFailContainer(pod *v1.Pod) (containername string, err error) {
+	status, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	if err != nil {
+		return "", fmt.Errorf("Unable to get status for pod %q: %v", format.Pod(pod), err)
+	}
+	for _, c := range pod.Spec.Containers {
+		cs := status.FindContainerStatusByName(c.Name)
+		if cs == nil || cs.State != kubecontainer.ContainerStateRunning {
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: Is successful termination permitted here?

File: pkg/kubelet/runonce.go [hunk 2]
```diff
@@ -156,3 +156,18 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
...
 }
+
+// getFailContainer returns failed container name for pod.
+func (kl *Kubelet) getFailContainer(pod *v1.Pod) (containername string, err error) {
+	status, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	if err != nil {
+		return "", fmt.Errorf("Unable to get status for pod %q: %v", format.Pod(pod), err)
+	}
+	for _, c := range pod.Spec.Containers {
+		cs := status.FindContainerStatusByName(c.Name)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: container status has a name, just iterate over those.

File: pkg/kubelet/runonce.go [hunk 3]
```diff
@@ -156,3 +156,18 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
...
+		if cs == nil || cs.State != kubecontainer.ContainerStateRunning {
+			return c.Name, nil
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: this should return a slice

**Comment by hanxiaoshuai** [comment]
cc @tallclair I have re pushed it, Thanks

**Comment by hanxiaoshuai** [comment]
cc @tallclair

File: pkg/kubelet/runonce.go [hunk 4]
```diff
@@ -91,8 +91,8 @@ func (kl *Kubelet) runOnce(pods []*v1.Pod, retryDelay time.Duration) (results []
 		res := <-ch
 		results = append(results, res)
 		if res.Err != nil {
-			// TODO(proppy): report which containers failed the pod.
-			glog.Infof("failed to start pod %q: %v", format.Pod(res.Pod), res.Err)
+			failcontainername, _ := kl.getFailContainer(res.Pod)
+			glog.Infof("failed to start container %v for pod %q: %v", failcontainername, format.Pod(res.Pod), res.Err)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: Nit: capitalize `Failed` while you're at it.

File: pkg/kubelet/runonce.go [hunk 5]
```diff
@@ -156,3 +156,19 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
 	}
 	return true
 }
+
+// getFailContainer returns failed container name for pod.
+func (kl *Kubelet) getFailContainer(pod *v1.Pod) ([]string, error) {
+	status, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	containername := make([]string, 0)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: nit: move this after the error check

File: pkg/kubelet/runonce.go [hunk 6]
```diff
@@ -156,3 +156,19 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
 	}
 	return true
 }
+
+// getFailContainer returns failed container name for pod.
+func (kl *Kubelet) getFailContainer(pod *v1.Pod) ([]string, error) {
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: nit: rename to `getFailedContainers`

File: pkg/kubelet/runonce.go [hunk 7]
```diff
@@ -156,3 +156,19 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
...
+	status, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	containername := make([]string, 0)
+	if err != nil {
+		return containername, fmt.Errorf("Unable to get status for pod %q: %v", format.Pod(pod), err)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: nit: lowercase error message

File: pkg/kubelet/runonce.go (hunk 5 above)
Comment (tallclair) [reply pkg/kubelet/runonce.go]: nit: rename to `containerNames`

File: pkg/kubelet/runonce.go [hunk 8]
```diff
@@ -156,3 +156,19 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
...
+	}
+	for _, c := range pod.Spec.Containers {
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: https://github.com/kubernetes/kubernetes/pull/59537#discussion_r168866448

> container status has a name, just iterate over those.

Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: @tallclair I think it has iterated in func FindContainerStatusByName below, if I misunderstand it ?

File: pkg/kubelet/runonce.go (hunk 4 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 5 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 6 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 7 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 8 above)
Comment (tallclair) [reply pkg/kubelet/runonce.go]: What I meant was:
```go
for _, cs := range status.ContainerStatuses {
	if cs.State != kubecontainer.ContainerStateRunning && cs.ExitCode != 0 {
		containerNames = append(containerNames, cs.Name)
	}
}
```

File: pkg/kubelet/runonce.go [hunk 9]
```diff
@@ -156,3 +160,19 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
 	}
 	return true
 }
+
+// getFailContainer returns failed container name for pod.
+func (kl *Kubelet) getFailContainers(pod *v1.Pod) ([]string, error) {
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: nit: s/Fail/Failed/

File: pkg/kubelet/runonce.go (hunk 8 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 9 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done @tallclair

File: pkg/kubelet/runonce.go [hunk 10]
```diff
@@ -91,8 +91,12 @@ func (kl *Kubelet) runOnce(pods []*v1.Pod, retryDelay time.Duration) (results []
 		res := <-ch
 		results = append(results, res)
 		if res.Err != nil {
-			// TODO(proppy): report which containers failed the pod.
-			glog.Infof("failed to start pod %q: %v", format.Pod(res.Pod), res.Err)
+			failcontainername, err := kl.getFailedContainers(res.Pod)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: `faliedContainerName`

File: pkg/kubelet/runonce.go [hunk 11]
```diff
@@ -91,8 +91,12 @@ func (kl *Kubelet) runOnce(pods []*v1.Pod, retryDelay time.Duration) (results []
...
+			if err != nil {
+				glog.Infof("failed to getFailContainers for pod %q: %v, error:%v", format.Pod(res.Pod), res.Err, err)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: Capitalize the log message: `Failed to getFailedContainers ...`

File: pkg/kubelet/runonce.go [hunk 12]
```diff
@@ -91,8 +91,12 @@ func (kl *Kubelet) runOnce(pods []*v1.Pod, retryDelay time.Duration) (results []
...
+			} else {
+				glog.Infof("failed to start container %v for pod %q: %v", failcontainername, format.Pod(res.Pod), res.Err)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: capitalize

File: pkg/kubelet/runonce.go [hunk 13]
```diff
@@ -156,3 +160,18 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
...
 	return true
 }
+
+// getFailedContainer returns failed container name for pod.
+func (kl *Kubelet) getFailedContainers(pod *v1.Pod) ([]string, error) {
+	status, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	if err != nil {
+		return []string{}, fmt.Errorf("unable to get status for pod %q: %v", format.Pod(pod), err)
+	}
+	containerNames := make([]string, 0)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: nit: prefer `var containerNames []string`

File: pkg/kubelet/runonce.go [hunk 14]
```diff
@@ -156,3 +160,18 @@ func (kl *Kubelet) isPodRunning(pod *v1.Pod, status *kubecontainer.PodStatus) bo
 	}
 	return true
 }
+
+// getFailedContainer returns failed container name for pod.
+func (kl *Kubelet) getFailedContainers(pod *v1.Pod) ([]string, error) {
+	status, err := kl.containerRuntime.GetPodStatus(pod.UID, pod.Name, pod.Namespace)
+	if err != nil {
+		return []string{}, fmt.Errorf("unable to get status for pod %q: %v", format.Pod(pod), err)
```
Comment (tallclair) [inline pkg/kubelet/runonce.go]: `return nil, ...`

File: pkg/kubelet/runonce.go (hunk 10 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 11 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 12 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 13 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: Done

File: pkg/kubelet/runonce.go (hunk 14 above)
Comment (hanxiaoshuai) [reply pkg/kubelet/runonce.go]: @tallclair Done Thanks

**Pull request merged**